### ISSUE TITLE
hs.webview freezes?

### ISSUE THREAD

**Issue by latenitefilms** [Hammerspoon/hammerspoon#2334]
Every now and again if I have a `hs.webview` that is being updated through code, whilst something else is happening in Hammerspoon, I get a freeze. I've never really looked into it, or worked out why it does it.

A quick look at the `hs.webview` code, and I've noticed there's a couple of:

```objc
if (theView.loading) [theView stopLoading] ;
while (theView.loading) {}
```

@asmagill - I wonder if it's a possibility of `theView.loading` never returning false, hence the freeze?

**Comment by latenitefilms** [comment]
Does this give you any clues?

```lua
Date/Time:       2020-04-03 23:56:31 +1100
End time:        2020-04-03 23:56:32 +1100
OS Version:      Mac OS X 10.14.6 (Build 18G3020)
Architecture:    x86_64h
Report Version:  28

Data Source:     Stackshots
Shared Cache:    0x4aa6000 7F556C80-8A46-36EC-A2EF-6592B7F1F89D
Shared Cache:    0 3BF98159-9727-306D-B435-377EFF981064

Command:         CommandPost
Path:            /Users/USER/*/CommandPost.app/Contents/MacOS/CommandPost
Identifier:      org.latenitefilms.CommandPost
Version:         1.0.7-beta.6 (5878)
Parent:          launchd [1]
PID:             94383

Event:           hang
Duration:        1.20s (process was unresponsive for 15 seconds before sampling)
Steps:           12 (100ms sampling interval)

Hardware model:  MacBookPro14,3
Active cpus:     8

Time Awake Since Boot: 340000s
Time Since Wake: 300000s

Fan speed:       2537 rpm

--------------------------------------------------
Timeline format: stacks are sorted chronologically
Use -i and -heavy to re-report with count sorting
--------------------------------------------------


Heaviest stack for the main thread of the target process:
  12  start + 1 (libdyld.dylib + 91093) [0x7fff5c3e33d5]
  12  NSApplicationMain + 777 (AppKit + 10952) [0x7fff2da35ac8]
  12  -[NSApplication run] + 699 (AppKit + 79240) [0x7fff2da46588]
  12  -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1361 (AppKit + 103531) [0x7fff2da4c46b]
  12  _DPSNextEvent + 1135 (AppKit + 108413) [0x7fff2da4d77d]
  12  _BlockUntilNextEventMatchingListInModeWithFilter + 64 (HIToolbox + 44150) [0x7fff2f6b5c76]
  12  ReceiveNextEventCommon + 603 (HIToolbox + 44773) [0x7fff2f6b5ee5]
  12  RunCurrentEventLoopInMode + 292 (HIToolbox + 45483) [0x7fff2f6b61ab]
  12  CFRunLoopRunSpecific + 455 (CoreFoundation + 239214) [0x7fff3045766e]
  12  __CFRunLoopRun + 2289 (CoreFoundation + 242097) [0x7fff304581b1]
  12  __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9 (CoreFoundation + 244359) [0x7fff30458a87]
  12  _dispatch_main_queue_callback_4CF + 1135 (libdispatch.dylib + 59021) [0x7fff5c3a168d]
  12  _dispatch_client_callout + 8 (libdispatch.dylib + 13885) [0x7fff5c39663d]
  12  _dispatch_call_block_and_release + 12 (libdispatch.dylib + 9720) [0x7fff5c3955f8]
  12  __29-[HSToolbar performCallback:]_block_invoke.99 + 656 (toolbar_internal.m:208,17 in toolbar_internal.so + 9881) [0x1147ea699]
  12  -[LuaSkin protectedCallAndError:nargs:nresults:] + 55 (Skin.m:343,19 in LuaSkin + 9619) [0x10f72d593]
  12  -[LuaSkin protectedCallAndTraceback:nresults:] + 218 (Skin.m:329,18 in LuaSkin + 9420) [0x10f72d4cc]
  12  lua_pcallk + 236 (lapi.c:969,14 in LuaSkin + 104804) [0x10f744964]
  12  luaD_pcall + 110 (ldo.c:156,3 in LuaSkin + 131604) [0x10f74b214]
  12  luai_objcttry + 28 (lobjectivec_exceptions.m:84,3 in LuaSkin + 47893) [0x10f736b15]
  12  luaD_callnoyield + 21 (ldo.c:526,3 in LuaSkin + 130327) [0x10f74ad17]
  12  luaD_call + 64 (ldo.c:516,5 in LuaSkin + 130262) [0x10f74acd6]
  12  luaV_execute + 708 (lvm.c:1134,13 in LuaSkin + 113944) [0x10f746d18]
  12  luaD_precall + 681 (ldo.c:451,11 in LuaSkin + 129896) [0x10f74ab68]
  5   webview_html + 329 (internal.m:1397,20 in internal.so + 43561) [0x113267a29]
  3   -[WKWebView isLoading] + 28 (WebKit + 1606358) [0x7fff3ecc22d6]


Process:         CommandPost [94383]
UUID:            FD1D2E36-2791-33C5-9BBD-7AE8AA1EC3BE
Path:            /Users/USER/*/CommandPost.app/Contents/MacOS/CommandPost
Architecture:    x86_64
Parent:          launchd [1]
UID:             501
Footprint:       228.79 MB
Start time:      2020-04-03 23:56:31 +1100
End time:        2020-04-03 23:56:32 +1100
Num samples:     12 (1-12)
CPU Time:        1.105s (4.1G cycles, 10.7G instructions, 0.38c/i)
Note:            Unresponsive for 15 seconds before sampling
Note:            2 idle work queue threads omitted

  Thread 0x70abd4           DispatchQueue 1           12 samples (1-12)         priority 47 (base 47)     cpu time 1.100s (4.0G cycles, 10.7G instructions, 0.38c/i)
  <process frontmost, thread QoS user interactive (requested user interactive), process unclamped, process received importance donation from WindowServer [228], IO tier 0>
  12  start + 1 (libdyld.dylib + 91093) [0x7fff5c3e33d5] 1-12
    12  NSApplicationMain + 777 (AppKit + 10952) [0x7fff2da35ac8] 1-12
      12  -[NSApplication run] + 699 (AppKit + 79240) [0x7fff2da46588] 1-12
        12  -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1361 (AppKit + 103531) [0x7fff2da4c46b] 1-12
          12  _DPSNextEvent + 1135 (AppKit + 108413) [0x7fff2da4d77d] 1-12
            12  _BlockUntilNextEventMatchingListInModeWithFilter + 64 (HIToolbox + 44150) [0x7fff2f6b5c76] 1-12
              12  ReceiveNextEventCommon + 603 (HIToolbox + 44773) [0x7fff2f6b5ee5] 1-12
                12  RunCurrentEventLoopInMode + 292 (HIToolbox + 45483) [0x7fff2f6b61ab] 1-12
                  12  CFRunLoopRunSpecific + 455 (CoreFoundation + 239214) [0x7fff3045766e] 1-12
                    12  __CFRunLoopRun + 2289 (CoreFoundation + 242097) [0x7fff304581b1] 1-12
                      12  __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9 (CoreFoundation + 244359) [0x7fff30458a87] 1-12
                        12  _dispatch_main_queue_callback_4CF + 1135 (libdispatch.dylib + 59021) [0x7fff5c3a168d] 1-12
                          12  _dispatch_client_callout + 8 (libdispatch.dylib + 13885) [0x7fff5c39663d] 1-12
                            12  _dispatch_call_block_and_release + 12 (libdispatch.dylib + 9720) [0x7fff5c3955f8] 1-12
                              12  __29-[HSToolbar performCallback:]_block_invoke.99 + 656 (toolbar_internal.m:208,17 in toolbar_internal.so + 9881) [0x1147ea699] 1-12
                                12  -[LuaSkin protectedCallAndError:nargs:nresults:] + 55 (Skin.m:343,19 in LuaSkin + 9619) [0x10f72d593] 1-12
                                  12  -[LuaSkin protectedCallAndTraceback:nresults:] + 218 (Skin.m:329,18 in LuaSkin + 9420) [0x10f72d4cc] 1-12
                                    12  lua_pcallk + 236 (lapi.c:969,14 in LuaSkin + 104804) [0x10f744964] 1-12
                                      12  luaD_pcall + 110 (ldo.c:156,3 in LuaSkin + 131604) [0x10f74b214] 1-12
                                        12  luai_objcttry + 28 (lobjectivec_exceptions.m:84,3 in LuaSkin + 47893) [0x10f736b15] 1-12
                                          12  luaD_callnoyield + 21 (ldo.c:526,3 in LuaSkin + 130327) [0x10f74ad17] 1-12
                                            12  luaD_call + 64 (ldo.c:516,5 in LuaSkin + 130262) [0x10f74acd6] 1-12
                                              12  luaV_execute + 708 (lvm.c:1134,13 in LuaSkin + 113944) [0x10f746d18] 1-12
                                                12  luaD_precall + 681 (ldo.c:451,11 in LuaSkin + 129896) [0x10f74ab68] 1-12
                                                  1   webview_html + 320 (internal.m:1397,20 in internal.so + 43552) [0x113267a20] (running) 1
                                                  1   webview_html + 329 (internal.m:1397,20 in internal.so + 43561) [0x113267a29] 2
                                                    1   -[WKWebView isLoading] + 28 (WebKit + 1606358) [0x7fff3ecc22d6] 2
                                                      1   WebKit::PageLoadState::isLoading() const + 4 (WebKit + 1852992) [0x7fff3ecfe640] (running) 2
                                                  1   objc_msgSend + 0 (libobjc.A.dylib + 26240) [0x7fff5ac06680] (running) 3
                                                  1   webview_html + 329 (internal.m:1397,20 in internal.so + 43561) [0x113267a29] 4
                                                    1   -[WKWebView isLoading] + 23 (WebKit + 1606353) [0x7fff3ecc22d1] (running) 4
                                                  1   webview_html + 320 (internal.m:1397,20 in internal.so + 43552) [0x113267a20] (running) 5
                                                  1   objc_msgSend + 46 (libobjc.A.dylib + 26286) [0x7fff5ac066ae] (running) 6
                                                  1   webview_html + 329 (internal.m:1397,20 in internal.so + 43561) [0x113267a29] 7
                                                    1   -[WKWebView isLoading] + 28 (WebKit + 1606358) [0x7fff3ecc22d6] (running) 7
                                                  1   webview_html + 329 (internal.m:1397,5 in internal.so + 43561) [0x113267a29] (running) 8
                                                  1   objc_msgSend + 46 (libobjc.A.dylib + 26286) [0x7fff5ac066ae] (running) 9
                                                  1   webview_html + 320 (internal.m:1397,20 in internal.so + 43552) [0x113267a20] (running) 10
                                                  2   webview_html + 329 (internal.m:1397,20 in internal.so + 43561) [0x113267a29] 11-12
                                                    2   -[WKWebView isLoading] + 28 (WebKit + 1606358) [0x7fff3ecc22d6] (running) 11-12
```

**Comment by asmagill** [comment]
Just to confirm, have you made any changes to the `internal.m` file for webview in CommandPost -- in my source, line 1397 is 
~~~objc
        WKNavigation *navID = [theView loadHTMLString:theHTML baseURL:[NSURL URLWithString:theBaseURL]] ;
~~~
Is there anything difficult about the html string being passed in (javascript loops that get entered immediately, etc.)?

I have an idea about wrapping the whole thing in a dispatch and using a timer to check `theView.loading` every few nanoseconds rather than use an empty `while` that might make it more responsive, but I'm a little concerned about what might happen if `hs.webview:html` got invoked twice in rapid succession before the first had actually completed, so I need to think about it a little bit first...

**Comment by latenitefilms** [comment]
> Just to confirm, have you made any changes to the internal.m file for webview in CommandPost -- in my source, line 1397 is

Oh, sorry, yeah, [looks like I have](https://github.com/CommandPost/CommandPost-App/blob/master/extensions/webview/internal.m#L1397):

```objc
while (theView.loading) {}
```

I think what's happening is that I have `hs.webview.toolbar` buttons that when pressed, change the HTML content (essentially all our UI's in CommandPost are done this way - see attached screenshots). The HTML content does have lots of JavaScript, and we are sending JavaScript messages back to Lua-land when the page finishes loading.

I just tried wrapping the functions that are triggered when pressing the `hs.webview.toolbar` buttons into a `hs.timer.doAfter(0, fn)`, and that SEEMS to have helped, but will let you know if I still see any freezes.

Stay safe!

**Comment by asmagill** [comment]
That does sound like a plausible way for things to get locked up, kind of like using a synchronous `hs.http.get` against a web server being hosted by Hammerspoon -- both are trying to do something on the main loop at the same time.

Ok, I'll try and implement a fix tomorrow -- the idea I have is to move the while loop into a block invoked by a timer -- if `theView.loading` is true, it will just reset the trigger and return to the main application; if it's false, then it will do the actual update and clear the block from the run loop. The only sticking point is making sure that any unfinished queued blocks are cleared if a new one is issued but I think a module level static variable may be sufficient to catch this.

I'll keep you posted...

**Comment by asmagill** [comment]
(and yes, that synchronous request example is one that bit me in the rear and took a while for me to understand... felt sheepish because it was obvious once I thought about it...)

**Comment by latenitefilms** [comment]
@asmagill - No rush or pressure, but just wondering... did you end up having a chance to look at this?

**Comment by asmagill** [comment]
No, sorry, got side tracked. I'll leave this open in my browser as a reminder and try to get to it in the next couple of days.

**Comment by latenitefilms** [comment]
Fixed in e535e22.
